src/chainstate/stacks/db/blocks.rs

```diff
@@ -745,29 +777,33 @@ impl StacksChainState {
             StacksChainState::make_block_dir(blocks_dir, consensus_hash, &block_header_hash)
                 .expect("FATAL: failed to create block directory");
 
+        // try make this thread-safe. It's okay if this block gets copied more than once; we
+        // only care that at least one copy survives for further analysis.
+        let random_bytes = thread_rng().gen::<[u8; 8]>();
+        let random_bytes_str = to_hex(&random_bytes);
+        let index_block_hash = StacksBlockId::new(consensus_hash, block_header_hash);
+        let mut invalid_path =
+            StacksChainState::get_index_block_pathbuf(blocks_dir, &index_block_hash);
+        invalid_path
+            .file_name()
+            .expect("FATAL: index block path did not have file name");
+        invalid_path.set_extension(&format!("invalid-{}", &random_bytes_str));
+
+        fs::copy(&block_path, &invalid_path).expect(&format!(
+            "FATAL: failed to copy '{}' to '{}'",
+            &block_path,
+            &invalid_path.to_string_lossy(),
+        ));
+
```

Comment:
Won't these changes make it possible that an already-freed-block is copied to the invalid block location?

Reply:
Yes, and that could happen in the code before as well.

Before, you could have a race as follows:

* Thread A calls `free_block()` and verifies that the size was greater than zero
* Thread B calls `free_block()`, verifies that the size was greater than 0, and then races ahead to copy and truncate the file
* Thread A wakes up and copies the zero-sized truncated file

I can just revert this, or at least add back the initial size check so we don't gratuitously make empty files.

Reply:
Reverted in 993c124